keps/sig-auth/3257-trust-anchor-sets/README.md

```diff
@@ -0,0 +1,1348 @@
+  to distribute trust anchors to workloads.
+
+Each of these approaches has downsides:
+
+* Creating a ConfigMap in every namespace requires granting over-broad
+  permissions to the responsible controller (create ConfigMaps in **all**
+  namespaces).
+
+* Static trust anchor sets embedded directly in configuration complicate the job
+  of rotating the backing CA's roots of trust.
+
+* CRDs work well, but are inappropriate for use by core APIs.
+
+Each of these approaches would be simplified by an in-tree mechanism for
+distributing trust anchor sets.
+
+### Goals
+
+* Provide a Kubernetes data type that holds X.509 trust anchors for use wherever
+  a set of X.509 trust anchors is needed.
+
+* Make it easy for signer implementations (certificates/v1 or otherwise) to
+  publish their relevant trust anchors for use within the cluster.
+
+* Make it easy for workloads that need access to the trust anchors that back a
+  particular certificates/v1 signer to access them within the container
+  filesystem.
+
+* Allow cluster operators to configure webhook and aggregated API server
+  backends using that use trust anchor objects instead of inline PEM data.
+
+### Non-Goals
+
+* Distribute a default set of "system" trust anchors.  (This is a natural
+  extension, though).
+
+* Handle trust anchors expressed in other forms than PEM-wrapped, DER-formatted
+  X.509.
+
+## Proposal
+
+This proposal is centered around a new cluster-scoped ClusterTrustBundle
+resource, initially in the certificates/v1alpha1 API group.  The
+ClusterTrustBundle object can be thought of as a specialized configmap tailored
+to the X.509 trust anchor use case.  Introducing a dedicated type allows us to
+attach different RBAC policies to ClusterTrustBundle objects, which will
+typically be wide-open for reading, but locked-down for writing.
+
+Each ClusterTrustBundle object is a container for multiple named trust anchor
+sets.  Most use cases will simply use a single trust anchor set named `default`,
+but mutual TLS use cases involving multiple federating meshes will require the
+ability to express multiple, non-commingled trust anchor sets.
+
+ClusterTrustBundles may optionally be associated with a signer using their
+`.spec.signerName` field.  ClusterTrustBundles that are associated with a signer
+have special handling of creates and updates, restricting mutating actions to
+those with an RBAC grant on the signer name (similar to the handling of
+CertificateSigningRequest objects).
+
+To enable consumption by workloads, the new `trustAnchors` Kubelet projected
+volume source allows projecting a trust anchor set from a particular signer into
+the container filesystem, with the contents of the projected files updating as
+the corresponding trust anchor sets are updated.
+
+In general, ClusterTrustBundle objects are considered publicly-readable within
+the cluster.  Read permissions will be granted cluster-wide to the
+`system:authenticated` group.
```

Comment:
Discussion at the last SIGAuth meeting brought up use-cases where cluster administrators would want to hide/make not enumerable the ClusterTrustBundle Objects, such as a shared cluster with namespace-per-customer where a ClusterTrustBundle was created for each customer. I got the impression that the consensus was that this was a valid use-case, so these permissions should be optional but recommended.

Reply:
This can't work, I think... if you can create a pod, you will be able to get the content of any ClusterTrustBundle.  We would need to add a namespace-scoped TrustBundle to support this usecase.

Reply:
I'm all for making them publically listable/readable and was just recording an objection I had heard. There may be room for difference between the "read" and "list" permissions, where you can get the contents of any trust bundle you know the name of, but not enumerate them; I don't think that makes sense and might lend a false sense of "security" to objects that should be considered public information.

Reply:
If the primary consumption method for these is a volume projection, why would we not just grant the permission to nodes? We don't grant workloads access to read config maps, etc... and we probably don't want to encourage random clients to get/list/watch these potentially very large objects.

In GKE, we would not be able to grant read to `system:authenticated`.

Reply:
Lets start with a default cluster role binding that grants read to `system:serviceaccounts` in the same we handled `system:service-account-issuer-discovery`.  Then users with `edit` level permissions can easily see this data via impersonation of any SA, and workloads that want to read the data directly can do so (i.e. in case they are running off cluster).

I consider unexpected users being granted access to GKE clusters via `system:authenticated` as a bug in how that environment handles authn, though I understand these things are hard to change once they exist.

Reply:
system:serviceaccounts seems less problematic, but it seems like all serviceaccounts would be covered by the Kubelet volume mounting ability...